cli/src/org/partiql/shell/ShellHighlighter.kt

```diff
@@ -13,130 +13,114 @@
-        }
-
-        // Check for Non-PartiQL Suffixes (REPL Only)
-        val lastNewlineIndex = lineIndexesMap[currentLine]!!
-        val suffixString = input.substring(lastNewlineIndex + 1, input.lastIndex + 1)
-        val lastValidQueryIndex = when (currentLine > 0 && ALLOWED_SUFFIXES.contains(suffixString)) {
-            true -> lastNewlineIndex
+        // Temporarily Remove Allowed Suffix from Input
+        val lastNewlineIndex = input.indexOfLast { c -> c == '\n' }
+        val suffixString = input.substring(lastNewlineIndex + 1)
+        val lastValidQueryIndex = when (lastNewlineIndex > 0 && ALLOWED_SUFFIXES.contains(suffixString)) {
+            true -> lastNewlineIndex + 1
             false -> input.length
         }
-
-        // Get Tokens
-        val tokens: List<Token>
-        try {
-            tokens = lexer.tokenize(input.substring(0, lastValidQueryIndex))
-        } catch (e: Exception) {
-            val error = AttributedString(input, AttributedStyle().foreground(AttributedStyle.RED))
-            return if (hasAddToGlobalEnv) {
-                with(AttributedStringBuilder()) {
-                    append(ADD_TO_GLOBAL_ENV_COMMAND)
-                    append(error)
-                    toAttributedString()
-                }
-            } else {
-                error
-            }
-        }
+        val usableInput = input.substring(0, lastValidQueryIndex)
 
         // Build Token Colors (Last Token is EOF)
+        val stream = getTokenStream(usableInput)
+        stream.fill()
+        val tokenIter = stream.tokens.iterator()
         var builder = AttributedStringBuilder()
-        for (tokenIndex in 0..(tokens.size - 2)) {
-            if (tokenIndex == tokens.lastIndex) break
-            val currentToken = tokens[tokenIndex]
-            val preIndex = when (tokenIndex) {
-                0 -> 0
-                else -> {
-                    val prevToken = tokens[tokenIndex - 1]
-                    (getTokenIndex(prevToken, lineIndexesMap) ?: 0) + prevToken.span.length.toInt()
+        var hasError = false
+        while (tokenIter.hasNext()) {
+            val token = tokenIter.next()
+            val type = token.type
+            val text = token.text
+            when {
+                isError(hasError, type) -> {
+                    hasError = true
+                    builder.styled(STYLE_ERROR, text.removeSuffix("<EOF>"))
                 }
+                isDatatype(type) -> builder.styled(STYLE_DATATYPE, text)
+                isIdentifier(type) -> builder.styled(STYLE_IDENTIFIER, text)
+                isString(type) -> builder.styled(STYLE_STRING, text)
+                isLiteral(type) -> builder.styled(STYLE_NUMBER, text)
+                isComment(type) -> builder.styled(STYLE_COMMENT, text)
+                isIonMode(type, text) -> builder.append(text)
+                isKeyword(type, text) -> builder.styled(STYLE_KEYWORD, text)
+                isEOF(type) -> { /* Do nothing */ }
```

Comment:
Is it possible (or necessary) to model those better? Like an Enum or a map?

Reply:
I actually prefer this way -- specifically because some of these functions are more than just `when` statements. Like, to remove verbosity, I opted to have a `when` statement for `isDatatype` which intentionally evaluates before `isKeyword`. And the reason for this is because `isKeyword` will actually take the input string and look into the PartiQLTokens vocabulary to see if it matches the keywords there (and datatypes) are keywords too.

Anyways, with this way, instead of having to write out every single keyword into an enum or map, I only need to write out the data types.

Reply:
No objection.